### ISSUE TITLE
Service worker cannot be enabled by function factory with AOT

### ISSUE THREAD

**Issue by Serginho** [angular/angular#22796]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ X ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
```ServiceWorkerModule.register('/ngsw-worker.js', {enabled: [boolean]});```

## Expected behavior
```ServiceWorkerModule.register('/ngsw-worker.js', {enabled: enableServiceWorker()});```

or

```{provide: RegistrationOptions, useFactory: () => {enabled: enableServiceWorker()}}```

This way is not possible because RegistrationOptions is not exported.

## What is the motivation / use case for changing the behavior?
Export RegistrationOptions provider in order to change the implementation.

The environment variable of my app is splitted in serveral files and joined with a merge function. So AOT fails registering service worker. The solution is enable it by Dependency injection but the provider is not public.
```export const environment = merge([{}, ENVIRONMENT_DEV, ENVIRONMENT_GLOBAL, { ... }]);```

## Environment

<pre><code>
Angular version: 5.2.9

**Comment by Serginho** [comment]
I have a provisional solution patching the module while the provider is private.

```
const modulePatch = ServiceWorkerModule.register('/ngsw-worker.js');
(<any>modulePatch.providers
  .find(provider =>
    (<any>provider).provide.toString().indexOf('RegistrationOptions'))
).useValue = enableServiceWorker();
```

**Comment by Serginho** [comment]
I have a provisional solution by patching the module while the providers is not exported.

```
const modulePatch = ServiceWorkerModule.register('/ngsw-worker.js');
(<any>modulePatch.providers
  .find(provider =>
    (<any>provider).provide.toString().match(/^function RegistrationOptions/))
).useValue = environment.production;
```

Any better solution?

**Comment by Serginho** [comment]
I think that enabled = true of RegistrationOptions is pretty unrealistic.

This doesn't work with AOT:
``` 
const serviceWorkerOptions = {enabled: environment.enableServiceWorker};
ServiceWorkerModule.register('/ngsw-worker.js', serviceWorkerOptions)
```

And this works, but ignore it!!!
```
const serviceWorkerOptions = {};
serviceWorkerOptions['enabled'] = environment.enableServiceWorker;
ServiceWorkerModule.register('/ngsw-worker.js', serviceWorkerOptions);
```

**Comment by ghetolay** [comment]
I'm having the same problem to not being able to use compile-time value for both the script and enabled option but this is not limited to service worker, it's a global AOT limitation. 
You can't call functions to compute metadatas, the same would happen for route definitions on RouterModule for example.

**Comment by kkirilov91** [comment]
There seems to be a feature addressing the problem:  https://github.com/angular/angular/pull/26002
Shouldn't be the issue closed?

**Comment by gkalpak** [comment]
Good catch, @kkirilov91. This has been fixed by #26002 indeed.
